### ISSUE TITLE
Hledger does not handle balance assertions when multiple commodities

### ISSUE THREAD

**Issue by thdox** [simonmichael/hledger#195]
Please see below:
- the test file
- works correctly with ledger (c++)
- fails with hledger

$ cat hledger_test_16

    2000/01/01 * Initial
        Actif:Courant:BnpCc                 -102411,12 F = -102411,12 F
        Actif:Courant:BnpCc                   21036,81 € = 21036,81 €
        Equity
     
    2002/01/11 * Passage à l'euro
        Actif:Courant:BnpCc                  102411,12 F = 0,00 F
        Actif:Courant:BnpCc                  -15612,47 € = 5424,34 €
        Revenu:Devise                       -102411,12 F
        Revenu:Devise                         15612,47 €

$ ledger --no-pager -f hledger_test_16 bal

               5424,34 €  Actif:Courant:BnpCc
             102411,12 F
             -21036,81 €  Equity
            -102411,12 F
              15612,47 €  Revenu:Devise
    --------------------
                       0

$ hledger -f hledger_test_16 bal

    hledger: Balance assertion failed for account Actif:Courant:BnpCc on 2000-01-01
    In transaction:
    2000/01/01 * Initial
        Actif:Courant:BnpCc  -102411,12 F
        Actif:Courant:BnpCc    21036,81 €
                              102411,12 F
        Equity                -21036,81 €
      
    After posting:
       Actif:Courant:BnpCc      21036,81 €  ;
     
    expected balance is 21036,81 €, actual balance was -102411,12 F
      21036,81 €.

**Comment by thdox** [comment]
Hi Simon, I've tested this against hledger 0.24.

It works correctly with test cases hledger_test_16a and hledger_test_16c (with an 'a' and a 'c' at the end of filename).
It does not work for hledger_test_16 (without letter at the end of filename).

I suspect this does not work because of your last comment "The ability to assert the full multi-commodity balance can be added again using the == syntax, if needed."
I do not fully catch this last comment. I went to http://hledger.org/manual.html, and a search can not find any occurrence of '=='.

I've tried syntax below, but this fails.

    2002/01/11 * Passage à l'euro
        Actif:Courant:BnpCc                  102411,12 F = 0,00 F
        Actif:Courant:BnpCc                  -15612,47 € == 5424,34 €
        Revenu:Devise                       -102411,12 F
        Revenu:Devise                         15612,47 €

Can you help me here by providing a syntax example that would fit with hledger_test_16 test case ?

Thanks

**Comment by thdox** [comment]
Hi Simon, 

There is one use case where I always use two balance assertions.
My broker, enable the possibility to sell an ETF, and buy a new one, instantly, without going through cash.
I note this, using ledger c++ syntax, as:

    2014/11/12 * Invest
        Actif:Invest:Broker    644,350 ETFone @@ 2000 €
        Actif:Invest   -2000 €
     
    2014/11/12 * Exchange
        Actif:Invest:Broker    -644,350 ETFone @@ 2197,23 €  = 0,000 ETFone  ; @ 3,4100 €
        Actif:Invest:Broker    30,587 ETFtwo @@ 2197,23 €  = 30,587 ETFtwo  ; @ 71,8340 €

Surprisingly, this works, I mean I can use two postings in one transaction, with two balance assertions

Thanks
Thierry

**Comment by simonmichael** [comment]
Hi Thierry,

`==` has not been implemented - I meant to suggest we could use that syntax if we want to provide the old (multi-commodity) assertions again.

Good catch with hledger_test_16 - something funny is going on there >:(. Here's a minimal example, commenting either the second posting or second transaction makes it work again:

1/1
    (a)   1F = 1F
    (a)   1G = 1G

1/2
    (a)   1F = 2F

**Comment by simonmichael** [comment]
Also thanks for the real-world examples, they are very useful. What's an ETF ?

You can have as many assertions as you have postings in a transaction.

I think you know this, but I'll just confirm these ways of writing your exchange transaction are all equivalent:
```
2014/11/12 * Exchange
    Actif:Invest:Broker    -644,350 ETFone @@ 2197,23 €  =  0,000 ETFone
    Actif:Invest:Broker      30,587 ETFtwo @@ 2197,23 €  = 30,587 ETFtwo

2014/11/12 * Exchange
    Actif:Invest:Broker    -644,350 ETFone @  3,4100 €  =  0,000 ETFone
    Actif:Invest:Broker      30,587 ETFtwo @ 71,8340 €  = 30,587 ETFtwo

2014/11/12 * Exchange
    Actif:Invest:Broker    -644,350 ETFone  =  0,000 ETFone
    Actif:Invest:Broker      30,587 ETFtwo  = 30,587 ETFtwo
```

**Comment by thdox** [comment]
Here is link to [wikipedia](http://en.wikipedia.org/wiki/Exchange-traded_fund) : basically, this is an investment fund, that holds stocks, with low fees.

I was surprised that transaction with ETF was working, as there are two balance assertions, and in hledger_test_16 file, there are also two balance assertions, but this fails. I can see that you re-open this issue #195, so I understand that you accept there is still something to correct. Thanks.

Thierry

**Comment by simonmichael** [comment]
Yes I will look at it. Thanks for the report.

**Comment by simonmichael** [comment]
Fixed by the 0.24.1 release.
